consensus/core/src/networktype.rs

```diff
@@ -104,3 +97,124 @@ impl Display for NetworkType {
         f.write_str(s)
     }
 }
+
+#[derive(thiserror::Error, PartialEq, Eq, Debug, Clone)]
+pub enum NetworkIdError {
+    #[error("Invalid network name prefix: {0}. The expected prefix is 'kaspa'.")]
+    InvalidPrefix(String),
+
+    #[error(transparent)]
+    InvalidNetworkType(#[from] NetworkTypeError),
+
+    #[error("Invalid network suffix: {0}. Only 32 bits unsigned integer (u32) are supported.")]
+    InvalidSuffix(String),
+}
+
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, BorshSerialize, BorshDeserialize, BorshSchema, PartialEq, Eq)]
+#[serde(rename_all = "lowercase")]
+pub struct NetworkId {
+    pub network_type: NetworkType,
+    pub suffix: Option<u32>,
+}
+
+impl NetworkId {
+    pub fn new(network_type: NetworkType, suffix: Option<u32>) -> Self {
+        Self { network_type, suffix }
+    }
+
+    pub fn name(&self) -> String {
+        if let Some(suffix) = self.suffix {
+            format!("kaspa-{}-{}", self.network_type, suffix)
+        } else {
+            format!("kaspa-{}", self.network_type)
+        }
+    }
+}
+
+impl Deref for NetworkId {
+    type Target = NetworkType;
+
+    fn deref(&self) -> &Self::Target {
+        &self.network_type
+    }
+}
+
+impl From<NetworkType> for NetworkId {
+    fn from(value: NetworkType) -> Self {
+        Self::new(value, None)
+    }
+}
+
+impl From<NetworkId> for Prefix {
+    fn from(net: NetworkId) -> Self {
+        (*net).into()
+    }
+}
+
+impl FromStr for NetworkId {
+    type Err = NetworkIdError;
+    fn from_str(network_name: &str) -> Result<Self, Self::Err> {
+        let mut parts = network_name.split('-');
+        let prefix = parts.next().unwrap_or_default();
+        if prefix != "kaspa" {
+            return Err(NetworkIdError::InvalidPrefix(prefix.to_string()));
+        }
+        let network_type = NetworkType::from_str(parts.next().unwrap_or_default())?;
+        let suffix = parts.next().map(|x| u32::from_str(x).map_err(|_| NetworkIdError::InvalidSuffix(x.to_string()))).transpose()?;
+        Ok(Self::new(network_type, suffix))
+    }
+}
+
+impl Display for NetworkId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.name())
```

Comment:
Ideally you implement the name logic here (w/o String alloc), and in `name()` you just call `to_string`